### PR TITLE
Fix bug in clientCodec.WriteRequest and serverCodec.WriteResponse.

### PR DESCRIPTION
**Pull request by petermattis**
WriteRequest and WriteResponse need to be safe for concurrent use. Both
were writing out the header and body of a request to the underlying conn
without proper locking which would allow interleaving of the data.

Discovered this as part of some network optimization work which is also
present in this change. Wrap bufio readers and writers around the
connection. Use the optimized marshalers and unmarshalers for the
request and response headers. Remove snappy compression which benchmarks
as a pure loss (presumably due to the Go implementation). Removed
checksumming of the body which isn't present in the other Go rpc codecs
and doesn't seem worthwhile given we have end-to-end checksums of user
data along with the TCP checksums. Reduced allocations in various
network code paths.

benchmark              old ns/op     new ns/op     delta
BenchmarkEchoGoRPC     45062         43120         -4.31%
BenchmarkEchoRPC       101950        38029         -62.70%

benchmark              old MB/s     new MB/s     speedup
BenchmarkEchoGoRPC     22.72        23.75        1.05x
BenchmarkEchoRPC       10.04        26.93        2.68x

### CODE REVIEW

**Comment by bdarnell** [comment]
What do you mean snappy is a pure loss? It costs CPU to be sure but the bandwidth savings will in theory make up for it when we're not just testing on localhost.

File: rpc/codec/conn.go [hunk 1]
```diff
@@ -5,92 +5,85 @@
...
+var framePool = sync.Pool{
+	New: func() interface{} {
+		return &frameBuf{}
+	},
+}
+
 func sendFrame(w io.Writer, data []byte) (err error) {
 	// Allocate enough space for the biggest uvarint
-	var size [binary.MaxVarintLen64]byte
+	buf := framePool.Get().(*frameBuf)
```
Comment (bdarnell) [inline rpc/codec/conn.go]: Why this change? Did the escape analysis decide that this array couldn't be put on the stack?

**Comment by petermattis** [comment]
@bdarnell I tested on 2 nodes in google compute engine and couldn't find a case where the compression helped performance (e.g. using compressible data which the current benchmarks are not doing). It's possible I did something wrong in the testing here. I was a bit surprised by this as well.

File: rpc/codec/rpc_test.go [hunk 2]
```diff
@@ -250,3 +253,93 @@ func testEchoClientAsync(t *testing.T, client *rpc.Client) {
 		)
 	}
 }
+
+func randString(n int) string {
```
Comment (spencerkimball) [inline rpc/codec/rpc_test.go]: Isn't this available via cockroach/util?

File: rpc/codec/conn.go (hunk 1 above)
Comment (petermattis) [inline rpc/codec/conn.go]: Yes, that is exactly what happened. I find this to be one of Go's soft spots.

**Comment by spencerkimball** [comment]
LGTM

File: rpc/codec/rpc_test.go (hunk 2 above)
Comment (petermattis) [inline rpc/codec/rpc_test.go]: Yes, but nothing in this package depends on `cockroach/util`. Thought it was better to minimize dependencies.

**Comment by bdarnell** [comment]
LGTM. Check in the benchmarks you used on GCE too so we can test things the same way when evaluating c-snappy.

File: rpc/codec/conn.go (hunk 1 above)
Comment (tschottdorf) [inline rpc/codec/conn.go]: Can you quickly explain/link how you found that out? I know about `-gcflags=-m` but that doesn't seem to answer it.

Comment (petermattis) [inline rpc/codec/conn.go]: I found this using the memory profiler: `go test -bench=. -memprofile=prof -memprofilerate=1`. The `-memprofilerate=1` setting tells the memory profiler to record every allocation. This slows the benchmark down a lot, but gives a price view of where the allocations are occurring. I then use `go tool pprof --alloc_objects <binary> prof` to view the output.

**Comment by petermattis** [comment]
@bdarnell my previous GCE test was one-off code that I didn't keep. I just reworked the Echo benchmarks so that they could be used between machines and replicated the previous behavior. With snappy disabled I achieved 434 MB/s. With snappy enabled I only achieved 280 MB/s. This benchmark was with 64KB compressible messages. I verified that snappy was achieving significant compression on the requests and responses (65540 -> 3142).
